simulation/modules/macrophage.py

```diff
@@ -109,51 +131,164 @@ def advance(self, state: State, previous_time: float):
+        #    next_mol_amount = iron[vox.z, vox.y, vox.x] ...
+
+
+def recruit(cells: MacrophageCellList, tissue, grid: RectangularGrid):
+    # TODO - add recruitment
+    # indices = np.argwhere(molecule_to_recruit >= threshold_value)
+    # then for each index create a cell with prob 'rec_rate'
+    return
+
+
+def remove(cells: MacrophageCellList, tissue, grid: RectangularGrid):
+    # TODO - add leaving
+    # indices = np.argwhere(molecule_to_leave <= threshold_value)
+    # then for each index kill a cell with prob 'leave_rate'
+    return
+
+
+def update(cells: MacrophageCellList, tissue, grid: RectangularGrid):
+    # TODO - add boolena network update
+    # for index in cells.alive:
+    #   cells[index].update_boolean network
+    return
+
+
+# move
+def chemotaxis(
+    molecule,
+    prob,
+    drift_lambda,
+    drift_bias,
+    cells: MacrophageCellList,
+    tissue,
+    grid: RectangularGrid,
+):
+    # 'molecule' = state.'molecule'.concentration
+    # prob = 0-1 random number to determine which voxel is chosen to move
+
+    # 1. Get cells that are alive
+    for index in cells.alive():
+
+        # 2. Get voxel for each cell to get molecule in that voxel
         cell = cells[index]
-        new_point = cell['point'] + delta_point
-        vox = grid.get_voxel(new_point)
-        while not (
-            valid_point(new_point, grid)
-            and tissue[vox.z, vox.y, vox.x] == TissueTypes.SURFACTANT.value
-        ):
-            x = random.uniform(-1, 1)
-            y = random.uniform(-1, 1)
-            z = random.uniform(-1, 1)
-            new_point = cell['point'] + Point(x=x, y=y, z=z)
-            vox = grid.get_voxel(new_point)
-
-        cell['point'] = new_point
-        cells.update_voxel_index([index])
-
-    return cells
+        vox = grid.get_voxel(cell['point'])
+
+        # 3. Set prob for neighboring voxels
+        p = []
+        vox_list = []
+        p_tot = 0.0
+        i = -1
+
+        # calculate individual probability
+        for x in [0, 1, -1]:
+            for y in [0, 1, -1]:
+                for z in [0, 1, -1]:
+                    p.append(0.0)
+                    vox_list.append([x, y, z])
+                    i += 1
+                    zk = vox.z + z
+                    yj = vox.y + y
+                    xi = vox.x + x
+                    if (
+                        (grid.xv[0] <= xi)
+                        & (xi <= grid.xv[-1])
+                        & (grid.yv[0] <= yj)
+                        & (yj <= grid.yv[-1])
+                        & (grid.zv[0] <= zk)
+                        & (zk <= grid.zv[-1])
+                    ):
+                        if tissue[zk, yj, xi] in [
+                            TissueTypes.SURFACTANT.value
+                            or TissueTypes.BLOOD.value
+                            or TissueTypes.EPITHELIUM.value
+                            or TissueTypes.PORE.value
+                        ]:

```

Comment:
Actually, this still isn't right.  You want a list here rather than logical "or".